[DON'T MERGE] Alternative hashing interface

lorentey opened the pull request:
This PR implements an alternative hashing interface that moves the choice of a hash function out of `Hashable`. Doing this allows us to evolve the hash function in the stdlib without recompiling hashable types.

To achieve this, it defines a streaming interface where hashable types can feed some stateful hash function with a variable number of `Int`s representing the bits to be hashed:

```
public protocol Hashable : Equatable {
  var hashValue: Int { get }
  func _hash(into hasher: (Int) -> Void)
}
```

The new `_hash(into:)` requirement comes with a default implementation for compatibility. However, it replaces `hashValue` as the primary hashing API for `Set` and `Dictionary`.

I expect we won't need to make the new method public. Outside the stdlib, it will be implemented by synthesized conformances of `Hashable`. Manually conforming `Hashable` types will continue to implement `hashValue`, as they have done since Swift 1.0. (We can't deprecate `hashValue` without significant source compatibility issues, and it doesn't seem worth the upgrade churn.)

lorentey committed:
[stdlib] Implement _hash(into:) in standard Hashable types.

lorentey committed:
[stdlib] Simplify SipHash13 and SipHash24 implementations

We don't need to stream bytes, just words.

lorentey committed:
[stdlib] Add alternative, streaming interface for hashing

- Add the method _hash(into:) as a Hashable requirement. It takes a closure into which the function can feed integer values representing bits to be hashed.
- Implement _hash(into:) in terms of hashValue in an extension, for compatibility.
- Add struct _QuickHasher, for emulating Swift 4 hashes in the new interface.
- Add _defaultHashValue<T>(for:), implementing hashValue in terms of _hash(into:) and _QuickHasher.

lorentey committed:
[stdlib] Update Dictionary and Set to use hash(into:) rather than hashValue

lorentey committed:
[WIP][SQUASH ME] Preparations for synthesizing Hashable with hashers

lorentey commented:
@swift-ci please benchmark

lorentey commented:
@swift-ci please smoke test

lorentey commented:
(Some hash values are different from master, so some tests will fail.)

lorentey commented:
cc @airspeedswift

lorentey commented:
@swift-ci please benchmark

lorentey committed:
[experiment] func _hash(into hasher: _Hasher) -> _Hasher

This is a version for hasher that uses a nonmutating append, so that we can mark it @effects(readonly), reducing retain/releases around its calls.

lorentey commented:
@swift-ci please benchmark

lorentey commented:
c433efd applies the lesson from #14466 directly onto the `_hash(into:)` interface, making the hash combinator a pure function, marked with `@effects(readonly)`.

This may speed things up a bit, but in exchange, we lose the ability to select a hasher at runtime, and the size of the hasher state likely becomes a limiting factor. (SipHash has a state that is 5 words wide.)

lorentey committed:
Add missing @_inlineable on default _hash(into:) implementation

This restores speed of String hashing, among other types.

lorentey commented:
@swift-ci please benchmark

lorentey commented:
Very good! We're down to +40% on synthetic microbenchmarks.

lorentey committed:
[stdlib] Switch to a faux-pure hasher interface

lorentey commented:
The latest commit should have equivalent or better performance, while hopefully also giving a boost to #14511's benchmarks. (This also enables us to remove the size of the hash state from the ABI.)

lorentey commented:
@swift-ci please benchmark

lorentey commented on stdlib/public/core/Hashable.swift:
```diff
@@ -109,20 +109,23 @@ public protocol Hashable : Equatable {
+  var hasher = _Hasher(_inlineable: ())
+  return withUnsafeMutablePointer(to: &hasher) { p in
+    return _UnsafeHasher(p).appending(value).finalized()
+  }
```
Cc @milseman

lorentey committed:
[benchmark] Update DictTest4 benchmark for new _hash(into:) variant

lorentey commented:
@swift-ci please benchmark

lorentey commented:
Wow, that's not how I expected this to go at all. Looks like we're not done yet.

I switched `finalize()` to an opaque call this time around; that may have contributed to the slowdown.

lorentey commented:
Frustratingly, I see quite reasonable +20% slowdowns in my local benchmarking runs; definitely nowhere near +452% for `SetIntersect`.

lorentey committed:
_UnsafeHasher.finalized(): Add @effects(readonly)

lorentey commented:
@swift-ci please benchmark

lorentey commented:
The benchmark is chugging along at https://ci.swift.org/job/swift-PR-osx-perf/141/

I expect the latest commit will get us back to acceptable performance.

milseman commented on stdlib/public/core/Hashable.swift:
```diff
@@ -185,10 +185,11 @@ public struct _UnsafeHasher {
 
-  @_inlineable
+  @effects(readonly) // See comment on appending(_:) above
+  @inline(never)
```
Why is this inline never? Shouldn't this be inlined into any part of the standard library that's behind a resilience barrier, such as the internals of Dictionary?

lorentey replied:
Indeed! This really only demonstrates how much effect a single additional non-inlineable function call during hashing has on our benchmarks. (By accident, it also highlighted the effect of `@effects(readonly)` rather nicely.) The benchmark below seems like a good target to shoot for.

The size of the `_Hasher` struct and its initializer are both baked into binaries at the moment. The entire `_Hasher` should be unversioned. Turns out, doing that by naively slapping `@inline(never)` on `_defaultHashValue` slows benchmarks down by up to 1800%.

milseman replied:
If we want to stack allocate the state on the stack, that makes it fixed layout. (Assuming non-alloca solutions here, but perhaps @aschwaighofer's improvements to stack-allocating generic/opaque values from last year would help).

When the user has defined their own `hashValue` on one member of a struct, and is deriving Hashable conformance for the whole struct, is the result of that member's hashValue fed into the `_Hasher` instead of the member's data? Can we partition hashing into a main `_Hasher` that's visiting everything, sometimes temporarily spawning a local `_Hasher` for any nested commutative collections before feeding that result back to the main one? If so, the main one can live on TLS and the embedded ones can be opaque/indirect. Then, we only pay the opaque layout cost once per nested commutative structure.

lorentey replied:
`_Hasher` can be stack-allocated as long as it’s initialized, fed and finalized inside a resilient function, right? It can have a fixed layout and fully inlineable methods as long as it’s not `@_versioned`.

But I’m not yet sure if there is a way to effectively pass a `Hashable`‘s `_hash(into:)` to such a function without incurring unacceptable overhead. Indeed, TLS for the root hasher will likely be the solution if we can’t.

lorentey committed:
Add _hash(into:) implementations for types conditionally conforming to Hashable

lorentey committed:
_defaultHashValue(for:) => _hashValue(for:)

lorentey committed:
Hide _Hasher entirely behind resilient function boundaries

lorentey commented:
So 543ee0c makes `_Hasher` entirely internal to the stdlib, hiding it behind resilience boundaries. Its initializer is now called in an `@inline(never)` function (`_hashValue(for:)`), so the hasher is stack-allocated -- however, since the function cannot be specialized for individual `Hashable` types, it needs to resort to `UnsafePointer` tricks to prevent 18x slowdowns due to boxing.

This will still have +500% slowdowns, though, I assume because of retain/release calls around `_hashValue(for:)`. Adding `@effects(readonly)` to `_hashValue(for:)` will eliminate these in a followup commit, but it's unclear to me if doing that is a good idea -- it potentially calls into user-provided `hashValue` implementations, which may do weird stuff.

lorentey commented:
@swift-ci benchmark

lorentey committed:
Make _hashValue(for:) public.

This is necessary for synthesized Hashable conformances.

lorentey committed:
Remove _hashValue(with:)

lorentey committed:
Add @effects(readonly) to non-inlineable version of _hashValue(for:).

lorentey commented:
@swift-ci please benchmark

lorentey committed:
First draft of synthesized Hashable conformance with _hash(into:)

This is rather tricky, as _hash(into:) has a default implementation and as such it wouldn't normally be synthesized. We piggyback on hashValue and unconditionally generate _hash(into:) as well.

lorentey committed:
[benchmark] DictTest4: Remove obsolete _hash(into:) definition

lorentey committed:
Synthesize _hash(into:) instead of providing a default implementation.

This *almost* works, except the compiler doesn't currently support deriving _hash(into:) method implementations for imported enums that are implicitly Hashable.

lorentey committed:
Cover all special cases of implicit Hashable conformance.

- Assume derived _hash(into:) typechecks for imported types
- Mark _hash(into:) as an external decl when the type itself was imported or synthesized

lorentey committed:
Review SipHash implementation and switch to SipHash13

lorentey committed:
Merge branch 'master' of https://github.com/apple/swift into smashing-hashing.synthesized

# Conflicts:
#	lib/Sema/DerivedConformanceEquatableHashable.cpp

lorentey commented:
@swift-ci please benchmark

lorentey commented:
This enables SipHash13 and synthesized _hash(into:) implementations.

lorentey committed:
Merge branch 'master' into smashing-hashing

lorentey commented:
@swift-ci please benchmark

lorentey committed:
FlattenCollection.Index.hashValue: Delegate to _hash(into:).

lorentey committed:
[stdlib] Move _UnsafeHasher and its posse to Hashing.swift.

lorentey committed:
[stdlib] Update tests that assumed specific hash values or Dictionary/Set ordering

lorentey committed:
Update tests looking at Hashable members

lorentey committed:
[Foundation] _CFObject: Add _hash(into:) implementation.

lorentey committed:
Remove test for _squeezeHashValue.

lorentey committed:
Revert to having a default implementation of _hash(into:)

Always synthesizing _hash(into:) almost worked, but it still had trouble with some imported classes and it had diagnostics regressions that seemed impossible to resolve without a major rework of how DerivedConformance works.

This approach works more reliably although it's a lot uglier.

lorentey committed:
Update tests looking at code completions in Hashable types.

lorentey committed:
StdlibUnittest: zero out the hash key in TestSuite.init.

This eliminates the chance of random test failures due to changing hash values, freak hash collision situations and unstable Dictionary ordering.

On the other hand, it allows tests to rely on specific hash values, which is unwise. It also requires that a TestSuite be initialized before the first element is inserted into any Set or Dictionary.

lorentey committed:
[DELETE] Temporarily disable SipHash tests.

lorentey commented:
@swift-ci please test

lorentey committed:
[stdlib] SipHash: Fix 32-bit platform support

Also, support appending 32-bit values on 64-bit platforms, too.

lorentey commented:
@swift-ci please test

lorentey committed:
Merge branch 'master' into smashing-hashing

lorentey committed:
[AST] Remove _combineHashValues and _mixInt from ASTContext

lorentey committed:
[stdlib] KeyPathComponent: Switch to new-style hashing

lorentey committed:
[SQUASH] Fix typo

lorentey committed:
[stdlib] Hash integers <= 32 bits as UInt32

lorentey committed:
[stdlib] SipHash: Revive ability to hash byte sequences of arbitrary length; restore some tests.

Appending is still limited to 32- or 64-bit values. However, a new version of finalize accepts a partial word to append at the end of the sequence. This is only used for testing.

lorentey commented:
A cleaned-up version of this lives on in #14913. See you there!
